### PR TITLE
fix(loki.write): Potential deadlock when queue is full and  component shuts down

### PR DESCRIPTION
**Pull request by kalleep**
### Brief description of Pull Request

Be able to shutdown in respect to drain timeouts when `loki.write` is blocked on full queue.

### Pull Request Details

Before this pr loki.write would not be able to perform updates or shutdown if any entry was blocked being enqueued. So the drain timeouts would not be respected.

To be able to handle both of these cases I had to remove channel usage from both of our consumer clients (WAL and non-WAL). IMO channel usage where the consumer of that channel is also responsible to close that channel is a bad pattern.

So instead I updated to just a plain function call `ConsumeEntry`. This one will propagate context from component so we can detect shutdown. We also "snapshot" consumer under read lock before calling `ConsumeEntry` so that we can unblock a potential update call. If update happens `loki.ErrConsumerStopped` will be returned and we have to abandon the entry. 

Alternative implementation to https://github.com/grafana/alloy/pull/7087

### Issue(s) fixed by this Pull Request

Related to: https://github.com/grafana/alloy/issues/7094

### Notes to the Reviewer

First commit in this pr adds 2 test that failed without the fix.

### PR Checklist

- [ ] Documentation added
- [x] Tests updated
- [ ] Config converters updated
- [ ] This pull request was substantially generated with AI assistance (see the [GenAI policy](https://github.com/grafana/alloy/blob/main/docs/developer/genai.md))

### CODE REVIEW

File: internal/component/common/loki/client/endpoint.go [hunk 1]
```diff
@@ -57,30 +43,34 @@ func newEndpoint(metrics *metrics, cfg Config, logger *slog.Logger, markerHandle
...
 
-		e.backoff.Wait()
-		if !e.backoff.Ongoing() {
-			return e.backoff.Err()
-		}
+		bo.Wait()
 	}
 
-	return nil
+	return bo.Err()
```
Comment (kgeckhart) [inline internal/component/common/loki/client/endpoint.go:75]: If we got here we dropped data due to a cancellation on the context breaking the bo, right?

**Review by kgeckhart — COMMENTED** [review]
Looks like both PRs still hang on a config update when the queue is full. Mine was supposed to fix it but it did not and goroutine dump on this one shows,

| Goroutine | Parked at | State |
|---|---|---|
| fanout `run` | `client/endpoint.go:67` (`bo.Wait()`) via `client/consumer_fanout.go:68` | retrying a full queue |
| mutator forwarder | `common/loki/entry_handler.go:86` | blocked on `nextChan <- f(e)` |
| `Component.Run` | `loki/write/write.go:139` | blocked on `c.sink.Chan() <- entry`, **holding `c.mut.RLock()`** |
| `Component.Update` | `loki/write/write.go:159` | blocked on `c.mut.Lock()` |

`Run`'s inner select only escapes on `ctx.Done()`, and a reload does not cancel the component context, so the read lock is never released.

File: internal/component/common/loki/client/endpoint.go (hunk 1 above)
Comment (kalleep) [reply internal/component/common/loki/client/endpoint.go:75]: Yeah so if WAL is enabled we dont since we dont update data as sent. But for non WAL we do drop data

File: internal/component/common/loki/client/consumer_fanout.go [hunk 2]
```diff
@@ -39,40 +40,37 @@ func NewFanoutConsumer(logger *slog.Logger, reg prometheus.Registerer, cfgs ...C
...
+		err := e.enqueue(ctx, entry, 0)
+
+		if err != nil {
+			// If we get errQueueIsFull we skipped the entry for this endpoints
+			// and should try next.
+			if errors.Is(err, errQueueIsFull) {
+				continue
+			}
+			// The other errors we can get are the context error and loki.ErrConsumerStopped.
+			// In both these cases there is no point trying to enqueue for other endpoints.
```
Comment (kgeckhart) [inline internal/component/common/loki/client/consumer_fanout.go]:
```suggestion
			// For any other error we assume it's not useful to send to the next endpoint
```

The comment about what errors are possible is liable to rot

File: internal/component/common/loki/client/consumer_wal.go [hunk 3]
```diff
@@ -108,8 +110,9 @@ type WALConsumer struct {
 	pairs  []endpointWatcherPair
 }
 
-func (m *WALConsumer) Chan() chan<- loki.Entry {
-	return m.writer.Chan()
+// ConsumeEntry implements DrainableConsumer.
+func (m *WALConsumer) ConsumeEntry(ctx context.Context, entry loki.Entry) error {
+	return m.writer.WriteEntry(entry)
 }
```
Comment (kgeckhart) [inline internal/component/common/loki/client/consumer_wal.go:116]: Is there any value is checking/propagating the context here or is something else going to cancel this during a drain/shutdown scenario?

File: internal/component/common/loki/client/endpoint.go [hunk 4]
```diff
@@ -35,52 +29,53 @@ func newEndpoint(metrics *metrics, cfg Config, logger *slog.Logger, markerHandle
...
+	for bo.Ongoing() {
+		err := e.shards.enqueue(tenantID, entry, segmentNum)
+
+		if err == nil {
+			return nil
+		}
+
+		if errors.Is(err, loki.ErrConsumerStopped) {
+			return err
+		}
```
Comment (kgeckhart) [inline internal/component/common/loki/client/endpoint.go:64]: Could we add a test on this one. It doesn't looks like it's currently tested and shutdown was another fun edge case on this one.

File: internal/component/common/loki/client/util_test.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+		case <-done:
+		case <-r.Context().Done():
+		}
+	}))
+
+	return server, blocked, release
+}
+
+func feedUntilBlocked(t *testing.T, blocked *atomic.Bool, consumer Consumer) {
+	e := loki.NewEntry(model.LabelSet{"A": "b"}, push.Entry{
```
Comment (kgeckhart) [inline internal/component/common/loki/client/util_test.go:43]: Nit
```suggestion
func feedUntilBlocked(t *testing.T, blocked *atomic.Bool, consumer Consumer) {
    t.Helper()
	e := loki.NewEntry(model.LabelSet{"A": "b"}, push.Entry{
```

Is this worth having a timeout on? There's nothing that couples it to newBlockedServer so something using it incorrectly would hang for a long time.

File: internal/component/common/loki/client/consumer_wal.go (hunk 3 above)
Comment (kalleep) [reply internal/component/common/loki/client/consumer_wal.go:116]: No not really, if we are in a drain state it's still fine to write entry to the WAL and it will be picked up when it's started again. If we have stopped it get back an error.

File: internal/component/common/loki/client/endpoint.go (hunk 4 above)
Comment (kalleep) [reply internal/component/common/loki/client/endpoint.go:64]: Yes will add one :+1:

Comment (kalleep) [reply internal/component/common/loki/client/endpoint.go:64]: added in [43ed462](https://github.com/grafana/alloy/pull/7091/commits/43ed462837693da97ccde1538ddfaddf44d930f6)

File: internal/component/common/loki/client/util_test.go (hunk 5 above)
Comment (kalleep) [reply internal/component/common/loki/client/util_test.go:43]: Sure I don't see why not, added a hard timeout in [840c140](https://github.com/grafana/alloy/pull/7091/commits/840c1407994680e5a32d04374ed250656b87f48b)

File: internal/component/loki/write/write.go [hunk 6]
```diff
@@ -199,22 +183,24 @@ func (c *Component) Update(args component.Arguments) error {
 	}
 
 	if err != nil {
-		return fmt.Errorf("failed to create cliens: %w", err)
+		return fmt.Errorf("failed to create clients: %w", err)
```
Comment (kgeckhart) [inline internal/component/loki/write/write.go:186]: For a follow up because it's an existing issue, AI called out that if creating the consumer fails `c.consumer` is going to be nil leading to a panic 😬 . Looks pretty legit since we do have some failure modes with empty/duplicate endpoints on consumer creation.

**Review by kgeckhart — APPROVED** [review]
Nice work!

**Review by clayton-cornell — APPROVED** [review]
Doc input is small.. no suggestions. Good as-is

File: internal/component/loki/write/write.go (hunk 6 above)
Comment (kalleep) [reply internal/component/loki/write/write.go:186]: Yes I saw that too. We also leak goroutines if we successfully create at least one endpoint but fails for another https://github.com/grafana/alloy/blob/main/internal/component/common/loki/client/consumer_fanout.go#L37.

Can fix both in another pr.
